core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -89,6 +92,11 @@
   private static final int BLOCK_REPLICATION_CONSTANT = 3;
   /** Lock for initializing the contexts, currently only one set of contexts is supported. */
   private static final Object INIT_LOCK = new Object();
+  /** Lock for cached contexts. */
+  private static final Object CACHED_CONTEXTS_LOCK = new Object();
+  /** Cache of FileSystem Contexts keyed on subjects. */
+  private static final Map<Subject, Pair<FileSystemContext, AtomicInteger>> CACHED_CONTEXTS =
```

Comment:
Do you think this cache could make more sense in its own class, or inside of `FileSystemContext`? Could this be used in non-hadoop contexts?

Reply:
We could move this to `FileSystemContext`, if we want to apply this change across the board, as opposed to only for access though the Hadoop API. WDYT?

Reply:
I'd like the caching to be consistent across all of our interfaces if possible. Do you see any issues with making the change across the board?